### DISCUSSION TITLE
Creating Extensions?

### DISCUSSION THREAD

**Discussion by TheOriginalBytePlayer** [sqlitebrowser/sqlitebrowser#3797] [Q&A]
Is it possible to for me to create custom extensions for SQLiteBrowser, and if so, what are the required parameters?

The reason I want to do this is that in my company's program, I am using SQLite3 less as a traditional database, and more as a storage mechanism for my program's project.  Because the data in the DB is (from the user's point of view) effectively one single document, I need to have the kinds of undo mechanisms one would expect for such a thing.  To that end, I have created a real-time automatic undo system through a series of OnBefore Triggers which all use a custom SQLiteFunction ShouldUndo() to know whether to store the undo or not. 

When I bring the project file into SQLiteBrowser, I have to first delete all the triggers before I can do anything as it doesn't understand the function ShouldUndo(). If I could create my own extension that supplied that function, it would make my life SO much easier.

So, can this be done, and where do I start if it can?

**Comment by justinclift** [comment]
Hmmmm.  I guess it depends if you're any good with C and or C++ programming?

SQLite itself is written in C, whereas our application is written in C++.  Depending upon your skill set it might be more feasible to approach this from a few different directions. :smile:

**Comment by TheOriginalBytePlayer** [comment]
I am a delphi programmer but I can write dlls and dylibs callable by C++ programs...

**Reply by justinclift** [reply]
Hmmm, not sure.  Maybe @mgrojo will have some suggestions or pointers?

**Reply by mgrojo** [reply]
You could write an [SQLite loadable extension](https://www.sqlite.org/loadext.html), but it has to export a C interface in order to be used by SQLite. If you can do that for C++, it should be even easier for C.

You can find examples for that extension in https://github.com/sqlitebrowser/sqlitebrowser/tree/master/src/extensions and in https://www.sqlite.org/src/dir?ci=tip&name=ext/misc
